### PR TITLE
Add support for TemplateInstance

### LINKED ISSUE #863: add TemplateInstance support
'oc get TemplateInstance' works in 3.6, but there is no way to do the same thing with the java client library.

### PR DESCRIPTION
**Pull request by kunal-kushwaha**
Fixes #863 #1380

### CODE REVIEW

**Comment by rohanKanojia** [comment]
ok to test

File: openshift-client/src/main/java/io/fabric8/openshift/client/dsl/internal/TemplateInstanceOperationsImpl.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+public class TemplateInstanceOperationsImpl
+  extends OpenShiftOperation<TemplateInstance, TemplateInstanceList, DoneableTemplateInstance, TemplateInstanceResource<TemplateInstance, KubernetesList, DoneableTemplateInstance>> {
+
+  public TemplateInstanceOperationsImpl(OkHttpClient client, OpenShiftConfig config) {
+    this(new OperationContext().withOkhttpClient(client).withConfig(config));
+  }
+
+  public TemplateInstanceOperationsImpl(OperationContext context) {
+    super(context.withApiGroupName("template.openshift.io")
```
Comment (gastaldi) [inline openshift-client/src/main/java/io/fabric8/openshift/client/dsl/internal/TemplateInstanceOperationsImpl.java]:
```suggestion
    super(context.withApiGroupName(OpenShiftAPIGroups.TEMPLATE)
```

File: openshift-client/src/main/java/io/fabric8/openshift/client/dsl/internal/TemplateInstanceOperationsImpl.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+package io.fabric8.openshift.client.dsl.internal;
+
+import io.fabric8.kubernetes.api.model.KubernetesList;
+import io.fabric8.kubernetes.client.dsl.base.OperationContext;
+import io.fabric8.openshift.api.model.DoneableTemplateInstance;
+import io.fabric8.openshift.api.model.TemplateInstance;
+import io.fabric8.openshift.api.model.TemplateInstanceList;
+import io.fabric8.openshift.client.OpenShiftConfig;
+import io.fabric8.openshift.client.dsl.TemplateInstanceResource;
```
Comment (gastaldi) [inline openshift-client/src/main/java/io/fabric8/openshift/client/dsl/internal/TemplateInstanceOperationsImpl.java]:
```suggestion
import io.fabric8.openshift.client.OpenShiftAPIGroups;
import io.fabric8.openshift.client.dsl.TemplateInstanceResource;
```

File: openshift-client/src/main/java/io/fabric8/openshift/client/DefaultOpenShiftClient.java [hunk 3]
```diff
@@ -31,6 +31,7 @@
 import io.fabric8.kubernetes.api.model.DoneableSecret;
 import io.fabric8.kubernetes.api.model.DoneableService;
 import io.fabric8.kubernetes.api.model.DoneableServiceAccount;
+import io.fabric8.kubernetes.api.model.Secret;
```
Comment (gastaldi) [inline openshift-client/src/main/java/io/fabric8/openshift/client/DefaultOpenShiftClient.java]:
```suggestion
```

**Comment by kunal-kushwaha** [comment]
@rohanKanojia: Can you please update the label for this PR?

**Comment by kunal-kushwaha** [comment]
@gastaldi: Requested changes made & commits sqaushed. Although, ci doesn't seem to be working.

**Comment by rohanKanojia** [comment]
retest this please

**Comment by rohanKanojia** [comment]
Were you facing these errors locally also?

**Comment by kunal-kushwaha** [comment]
@rohanKanojia: I just ran it again and this error showed up after I rebased with master branch to modify CHANGELOG.md
`[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project kubernetes-model: Fatal error compiling: java.lang.OutOfMemoryError: GC overhead limit exceeded -> [Help 1]`

**Comment by rohanKanojia** [comment]
oh, that's not good . We need to fix this

**Comment by kunal-kushwaha** [comment]
@rohanKanojia: I will look into what went wrong, any idea what might be causing this?

**Comment by rohanKanojia** [comment]
Try using `git bisect` to find out which commit is breaking this, We can revert it if it's possible.

**Comment by rohanKanojia** [comment]
I feel this might be due to some dependency upgrade

**Comment by gastaldi** [comment]
@kunal-kushwaha don't forget to squash all your commits into a single one in this PR. :wink:

File: extensions/knative/annotator/pom.xml [hunk 4]
```diff
@@ -53,6 +53,12 @@
         <dependency>
             <groupId>io.sundr</groupId>
             <artifactId>builder-annotations</artifactId>
+          <exclusions>
```
Comment (gastaldi) [inline extensions/knative/annotator/pom.xml]: How is this related to the TemplateInstance support?

Comment (kunal-kushwaha) [reply extensions/knative/annotator/pom.xml]: It was giving dependency issues for com.sun.tools:1.7 said that it's not available so just excluded it

Comment (gastaldi) [reply extensions/knative/annotator/pom.xml]: Weird, is your JAVA_HOME pointing to a JDK? Usually it fails when pointing to a JRE

Comment (kunal-kushwaha) [reply extensions/knative/annotator/pom.xml]: It's pointing to a JDK. This wasn't why the build was failing, it was just giving a warning

**Comment by rohanKanojia** [comment]
retest this please

**Comment by kunal-kushwaha** [comment]
@rohanKanojia: Thank You, I really appreciate it :-D

**Comment by kunal-kushwaha** [comment]
@gastaldi: Requested changes made.

**Comment by gastaldi** [comment]
[merge]
